airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/TypeAndDedupeOperationValve.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.base.destination.typing_deduping;
+
+import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * A slightly more complicated way to keep track of when to perform type and dedupe operations per
+ * stream
+ */
+public class TypeAndDedupeOperationValve extends ConcurrentHashMap<AirbyteStreamNameNamespacePair, Long> {
+
+  private static final long TWO_MINUTES_MILLIS = 1000 * 60 * 2;
+
+  private static final long FIVE_MINUTES_MILLIS = 1000 * 60 * 5;
+
+  private static final long TEN_MINUTES_MILLIS = 1000 * 60 * 10;
+
+  // 15 minutes is the maximum amount of time allowed between checkpoints as defined by
+  // The Airbyte Protocol
+  private static final long FIFTEEN_MINUTES_MILLIS = 1000 * 60 * 15;
+
+  // New users of airbyte likely want to see data flowing into their tables as soon as possible
+  // However, as their destination tables grow in size, typing and de-duping data becomes an expensive
+  // operation
+  // To strike a balance between showing data quickly and not slowing down the entire sync, we use an
+  // increasing
+  // interval based approach. This is not fancy, just hard coded intervals.
+  private static final List<Long> typeAndDedupeIncreasingIntervals = List.of(
+      TWO_MINUTES_MILLIS,
+      FIVE_MINUTES_MILLIS,
+      TEN_MINUTES_MILLIS,
+      FIFTEEN_MINUTES_MILLIS);
+
+  // Constantly getting the system time adds a bit of overhead, adding a minimum record count
+  // To reduce calls for system time
+  private static final int MINIMUM_RECORD_INTERVAL = 100;
+
+  private static final Supplier<Long> SYSTEM_NOW = () -> System.currentTimeMillis();
+
+  private ConcurrentHashMap<AirbyteStreamNameNamespacePair, Integer> incrementalIndex;
+
+  private final Supplier<Long> nowness;
+  private ConcurrentHashMap<AirbyteStreamNameNamespacePair, Long> recordCounts;
+
+  public TypeAndDedupeOperationValve() {
```

Comment:
does it work to have this accept a parsedcatalog? I.e. to prefill all the maps, instead of checking containsKey everywhere. And then I think you can delete `addStream` entirely